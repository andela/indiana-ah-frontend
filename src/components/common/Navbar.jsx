import React, { Fragment, Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { Navbar, Nav } from 'react-bootstrap';
import SignupContainer from '../SignupFormContainer.jsx';
import Modal from './Modal.jsx';
import { ProfileImg, ImageLogo } from '../../styles/styledComponents/Navigation.jsx';
import {
  customLogo,
  searchIcon,
  notificationIcon,
  profile,
  dropDownIcon
} from '../../assets/images/svg';
import Button from '../../styles/styledComponents/Button.jsx';
import InputField from './input/InputComponent.jsx';
import { signOutUser } from '../../redux/actions/authActions';

export class NavBar extends Component {
  state = {
    dropDown: false,
    modalIsOpen: false
  };

  openModal = () => {
    this.setState(() => ({ modalIsOpen: true }));
  };

  closeModal = () => {
    const {
      auth: { isLoading }
    } = this.props;
    if (!isLoading) this.setState(() => ({ modalIsOpen: false }));
  };

  displayForm = (form) => {
    this.setState({ modalIsOpen: true, modalContent: form });
  };

  componentWillUnmount() {
    document.removeEventListener('click', this.closeDropDown);
  }

  dropDown = (event) => {
    event.preventDefault();

    this.setState({ dropDown: true }, () => {
      document.addEventListener('click', this.closeDropDown);
    });
  };

  closeDropDown = () => {
    this.setState({ dropDown: false }, () => {
      document.removeEventListener('click', this.closeDropDown);
    });
  };

  render() {
    const { user, auth } = this.props;
    const userLInk = (
      <Nav>
        <img
          className="ml-auto d-none d-md-none d-xs-none d-lg-block"
          src={notificationIcon}
        />
        <ProfileImg
          src={profile}
          className="ml-5 mt-2 d-none d-md-none d-xs-none d-lg-block"
          alt="logo"
        />
        <div to="/signup" className="d-flex ft-size-2 ml-5
          align-items-center text-capitalize">
          {user.userData.username || user.userData.email}
          <img
            className="mx-3 d-none d-md-none d-lg-block d-xs-none"
            src={dropDownIcon}
            onClick={this.dropDown}
          />
          {this.state.dropDown && (
            <div className="dropDown-menu">
              <Link
                to="/"
                onClick={() => {
                  this.props.signOutUser();
                }}
              >
                Logout
              </Link>
            </div>
          )}
        </div>
      </Nav>
    );
    return (
      <Fragment>
        <Navbar bg="white shadow-sm px-5" expand="lg">
          <Navbar.Brand>
            <Link to="/">
              <ImageLogo>
                <img src={customLogo} className="feather" />
              </ImageLogo>
            </Link>
          </Navbar.Brand>
          <Navbar.Toggle aria-controls="basic-navbar-nav" />
          <Navbar.Collapse id="basic-navbar-nav">
            <Nav className="mr-auto ml-5">
              <InputField
                type="text"
                id="search"
                value=""
                placeholder="search"
                handleChange={() => {}}
                width="38rem"
              />
              <img className="search-logo" src={searchIcon} />
            </Nav>
            {auth.isAuthenticated ? (
              userLInk
            ) : (
              <Button onClick={() => this.displayForm('login')} className="ml-auto" sm>
                Login
              </Button>
            )}
          </Navbar.Collapse>
        </Navbar>
        <Modal
          modalIsOpen={this.state.modalIsOpen}
          closeModal={this.closeModal}
          body={
            this.state.modalContent === 'login' ? (
              <div>
                <h2>Login Form</h2>
              </div>
            ) : (
              <SignupContainer
                displayForm={this.displayForm}
                closeModal={this.closeModal}
              />
            )
          }
        />
      </Fragment>
    );
  }
}

NavBar.propTypes = {
  user: PropTypes.object,
  signOutUser: PropTypes.func,
  auth: PropTypes.object
};

export const mapStateToProps = state => ({
  user: state.user,
  auth: state.auth
});

export default connect(
  mapStateToProps,
  { signOutUser }
)(NavBar);
