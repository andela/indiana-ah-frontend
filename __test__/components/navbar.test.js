import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { mount, shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
import Navbar, { NavBar } from '../../src/components/common/Navbar.jsx';

const initialState = {
  auth: { isAuthenticated: false },
  user: { userData: { username: 'ozone4real' } }
};

const mockStore = configureStore([thunk]);
const store = mockStore(initialState);
const event = {
  preventDefault: jest.fn()
};

describe('Custom Navbar component test', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <MemoryRouter>
            <Navbar />
          </MemoryRouter>
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('renders a navbar', () => {
    const navbar = mount(
      <Provider store={store}>
        <MemoryRouter>
          <Navbar />
        </MemoryRouter>
      </Provider>
    );
    expect(navbar.find('.navbar')).toBeDefined();
    const inputField = navbar.find('input');
    const button = navbar.find('button');
    const button2 = navbar.find('button').at(1);
    button2.simulate('click');
    expect(inputField.length).toEqual(1);
    expect(button.length).toEqual(2);
  });

  it('tests for its functionalities', () => {
    const wrapper = shallow(<NavBar {...initialState} signOutUser={jest.fn()} />);
    wrapper.instance().openModal();
    expect(wrapper.state().modalIsOpen).toEqual(true);
    wrapper.instance().closeModal();
    expect(wrapper.state().modalIsOpen).toEqual(false);
    wrapper.instance().displayForm('login');
    wrapper.instance().dropDown(event);
    wrapper.instance().closeDropDown();
    wrapper.instance().componentWillUnmount();
    wrapper.setProps({ auth: { isAuthenticated: true } });
    wrapper.setState({ dropDown: true });
    wrapper
      .find('Link')
      .at(1)
      .simulate('click');
  });
});
