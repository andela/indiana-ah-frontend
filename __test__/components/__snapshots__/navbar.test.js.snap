// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom Navbar component test renders correctly 1`] = `
Array [
  <nav
    className="navbar navbar-expand-lg navbar-light bg-white shadow-sm px-5"
  >
    <span
      className="navbar-brand"
    >
      <a
        href="/"
        onClick={[Function]}
      >
        <div
          className="sc-bZQynM hVJJCu"
        >
          <img
            className="feather"
            src="test-file-stub"
          />
        </div>
      </a>
    </span>
    <button
      aria-controls="basic-navbar-nav"
      aria-label="Toggle navigation"
      className="navbar-toggler"
      onClick={[Function]}
      type="button"
    >
      <span
        className="navbar-toggler-icon"
      />
    </button>
    <div
      aria-expanded={null}
      className="navbar-collapse collapse"
      id="basic-navbar-nav"
    >
      <div
        className="mr-auto ml-5 navbar-nav"
        onKeyDown={[Function]}
      >
        <div
          className="input-field"
        >
          <input
            className="sc-bdVaJa OWCEx"
            id="search"
            onChange={[Function]}
            placeholder="search"
            required={true}
            type="text"
            value=""
          />
          <small
            className="text-danger ml-3 form-text"
          />
        </div>
        <img
          className="search-logo"
          src="test-file-stub"
        />
      </div>
      <button
        className="ml-auto sc-htpNat gbVmAb"
        onClick={[Function]}
      >
        Login
      </button>
    </div>
  </nav>,
  <div />,
]
`;
